python/ql/src/experimental/semmle/python/frameworks/XML.qll

```diff
@@ -0,0 +1,196 @@
+  private class XMLEtreeParser extends DataFlow::CallCfgNode, XMLParser::Range {
+    XMLEtreeParser() { this = xmlEtree().getMember("XMLParser").getACall() }
+
+    override DataFlow::Node getAnInput() { none() }
+
+    override predicate mayBeDangerous() { any() }
+  }
+
+  /**
+   * Gets a call to `xml.etree.ElementTree.fromstring`, `xml.etree.ElementTree.fromstringlist`,
+   * `xml.etree.ElementTree.XML` or `xml.etree.ElementTree.parse`.
+   *
+   * Given the following example:
+   *
+   * ```py
+   * parser = lxml.etree.XMLParser()
+   * parsed_xml = xml.etree.ElementTree.fromstring(xml_content, parser=parser).text
+   * ```
+   *
+   * `this` would be `xml.etree.ElementTree.fromstring(xml_content, parser=parser)`
+   * and `xml_content` would be the result of `getAnInput()`.
+   */
+  private class XMLEtreeParsing extends DataFlow::CallCfgNode, XMLParsing::Range {
+    XMLEtreeParsing() {
+      this = xmlEtree().getMember(["fromstring", "fromstringlist", "XML", "parse"]).getACall()
+    }
+
+    override DataFlow::Node getAnInput() { result = this.getArg(0) }
+
+    override predicate mayBeDangerous() {
+      exists(XMLParser xmlParser |
+        xmlParser.mayBeDangerous() and this.getArgByName("parser").getALocalSource() = xmlParser
+      )
+    }
+  }
+
+  /** Gets a reference to `xml.sax`. */
+  private API::Node xmlSax() { result = API::moduleImport("xml").getMember("sax") }
+
+  /**
+   * Gets a call to `xml.sax.make_parser` and following calls.
+   *
+   * Given the following example:
+   *
+   * ```py
+   * BadHandler = MainHandler()
+   * parser = xml.sax.make_parser()
+   * parser.setContentHandler(BadHandler)
+   * parser.setFeature(xml.sax.handler.feature_external_ges, False)
+   * parser.parse(StringIO(xml_content))
+   * parsed_xml = BadHandler._result
+   * ```
+   *
+   * `this` would be `xml.sax.make_parser()`, `getAnInput()` would return `StringIO(xml_content)`
+   * and `mayBeDangerous()` would not hold since `xml.sax.handler.feature_external_ges` is set to
+   * `False` and so is not vulnerable.
+   * see https://docs.python.org/3/library/xml.sax.handler.html#xml.sax.handler.feature_external_ges
+   */
+  private class XMLSaxParser extends DataFlow::CallCfgNode, XMLParser::Range {
+    DataFlow::CallCfgNode attrCall;
+
+    XMLSaxParser() {
+      this = xmlSax().getMember("make_parser").getACall() and
+      attrCall.getFunction().(DataFlow::AttrRead).getObject().getALocalSource() = this
+    }
+
+    override DataFlow::Node getAnInput() {
+      attrCall.getFunction().(DataFlow::AttrRead).getAttributeName() = "parse" and
+      result = attrCall.getArg(0)
+    }
+
+    override predicate mayBeDangerous() {
+      attrCall.getFunction().(DataFlow::AttrRead).getAttributeName() = "setFeature" and
+      attrCall.getArg(0) = xmlSax().getMember("handler").getMember("feature_external_ges").getAUse() and
+      DataFlow::localFlow(DataFlow::exprNode(any(True trueName)), attrCall.getArg(1))
+    }
+  }
```

Comment:
The way this code ends up being used in 

```
exists(XMLParser xmlParser | sink = xmlParser.getAnInput() and xmlParser.mayBeDangerous())
```

means that you might end up with some false-positives such as this contrived example

```
def contrived_example(user_input, action):
    parser = xml.sax.make_parser()
    if action == 'load-config':
        parser.setFeature(xml.sax.handler.feature_external_ges, False)
        parser.parse("/not-user-controlled/default_config.xml")
    else:
        parser.parse(StringIO(user_input))
```

since the second `parser.parse` call (with `user_input`) can be tracked back to the same parser that is used in some `setFeature` call. Instead, I would recommend using a forwards type-tracker to track XML SAX parsers that have the vulnerable feature enabled instead. (and starting off by adding this example as a test)